gen/enum.go

```diff
@@ -101,6 +106,96 @@ func enum(g Generator, spec *compile.EnumSpec) error {
 			<end>
 			return fmt.Sprintf("<$enumName>(%d)", <$w>)
 		}
+
+		func (<$v> <$enumName>) MarshalJSON() ([]byte, error) {
+			<if len .Spec.Items>
+				switch int32(<$v>) {
+				<range .UniqueItems>
+					case <.Value>:
+						return ([]byte)("\"<.Name>\""), nil
+				<end>
+				}
+			<end>
+			return ([]byte)(<$strconv>.FormatInt(int64(<$v>), 10)), nil
+		}
+
+		<$text := newVar "text">
+		func (<$v> *<$enumName>) UnmarshalJSON(<$text> []byte) error {
+			<$d := newVar "d">
+			<$t := newVar "t">
+
+			<$d> := <$json>.NewDecoder(<$bytes>.NewReader(<$text>))
+			<$d>.UseNumber()
+			<$t>, err := <$d>.Token()
+			if err != nil {
+				return err
+			}
+
+			<$ok := newVar "ok">
+			if <$w>, <$ok> := <$t>.(<$json>.Number); <$ok> {
+				<$x := newVar "x">
+				<$x>, err := <$w>.Int64()
+				if err != nil {
+					return err
+				}
+				if <$x> <">="> 0x80000000 {
+					return <$fmt>.Errorf("enum overflow from JSON %q for %q", <$text>, "<$enumName>")
+				}
+				if <$x> <"<"> -0x80000000 {
+					return <$fmt>.Errorf("enum underflow from JSON %q for %q", <$text>, "<$enumName>")
+				}
+				*<$v> = (<$enumName>)(<$x>)
+				return nil
+			}
+
+			if <$w>, <$ok> := <$t>.(string); <$ok> {
+				switch string(<$w>) {
+				<$enum := .Spec>
+				<range .Spec.Items>
+					case "<.Name>":
+						*<$v> = <enumItemName $enum .Name>
+						return nil
+				<end>
+					default:
+						return <$fmt>.Errorf("unknown enum value %q for %q", <$w>, "<$enumName>")
+				}
+			}
+
+			return <$fmt>.Errorf("invalid JSON value %q (%T) to unmarshal into %q", <$t>, <$t>, "<$enumName>")
+		}
+
+		func (<$v> *<$enumName>) UnmarshalJSON2(<$text> []byte) error {
```

Comment:
not used, and it was another way to implement it. one without using the json library, and likely faster because of that. both version worked with my unit tests the same. Anyway, we decided to go for the json version with abhinav, and I forgot to remove the code. It is now done.